agent/eventhandler/attachment_handler.go

```diff
@@ -0,0 +1,172 @@
+	lock sync.Mutex
+
+	client api.ECSClient
+	ctx    context.Context
+}
+
+// NewAttachmentEventHandler returns a new AttachmentEventHandler object
+func NewAttachmentEventHandler(ctx context.Context,
+	stateSaver statemanager.Saver,
+	client api.ECSClient) *AttachmentEventHandler {
+	return &AttachmentEventHandler{
+		ctx:                    ctx,
+		stateSaver:             stateSaver,
+		client:                 client,
+		attachmentARNToHandler: make(map[string]*attachmentHandler),
+		backoff: retry.NewExponentialBackoff(submitStateBackoffMin, submitStateBackoffMax,
+			submitStateBackoffJitterMultiple, submitStateBackoffMultiple),
+	}
+}
+
+// AddStateChangeEvent adds a state change event to AttachmentEventHandler for it to handle
+func (eventHandler *AttachmentEventHandler) AddStateChangeEvent(change statechange.Event) error {
+	if change.GetEventType() != statechange.AttachmentEvent {
+		return fmt.Errorf("eventhandler: attachment handler received unrecognized event type: %d", change.GetEventType())
+	}
+	event, ok := change.(api.AttachmentStateChange)
+	if !ok {
+		return errors.New("eventhandler: unable to get attachment event from state change event")
+	}
+
+	if event.Attachment == nil {
+		return fmt.Errorf("eventhandler: received malformed attachment state change event: %v", event)
+	}
+
+	attachmentARN := event.Attachment.AttachmentARN
+	eventHandler.lock.Lock()
+	if _, ok := eventHandler.attachmentARNToHandler[attachmentARN]; !ok {
+		eventHandler.attachmentARNToHandler[attachmentARN] = &attachmentHandler{
+			attachmentARN: attachmentARN,
+			stateSaver:    eventHandler.stateSaver,
+			client:        eventHandler.client,
+			ctx:           eventHandler.ctx,
+			backoff:       eventHandler.backoff,
+		}
+	}
+	eventHandler.lock.Unlock()
+
+	attachmentHandler := eventHandler.attachmentARNToHandler[attachmentARN]
+	go attachmentHandler.submitAttachmentEvent(&event)
+
+	return nil
+}
+
+// submitAttachmentEvent submits an attachment event to backend
+func (handler *attachmentHandler) submitAttachmentEvent(attachmentChange *api.AttachmentStateChange) {
+	// we need to lock the attachment handler to avoid sending an attachment state change for an attachment
+	// multiple times (this can happen when udev watch sends multiple attached events for a certain attachment,
+	// for example one from udev event and one from reconciliation loop)
+	seelog.Debugf("AttachmentHandler: acquiring attachment lock before sending attachment state change for attachment %s", handler.attachmentARN)
+	handler.lock.Lock()
+	seelog.Debugf("AttachmentHandler: acquired attachment lock for attachment %s", handler.attachmentARN)
+	defer handler.lock.Unlock()
+
+	// we need to continue retrying but skip submitting attachment state change when we successfully submitted the state
+	// change but failed to save state
+	retry.RetryWithBackoffCtx(handler.ctx, handler.backoff, func() error {
```

Comment:
There are far too many anonymous functions in this block. Please make it a point to come back and rectify that (with named methods) in the future (by creating an issue).

Reply:
added a TODO for this

Reply:
Can you please an issue and add a link to that as well, along with the `TODO` ?

Reply:
created the issue [here](https://github.com/aws/amazon-ecs-agent/issues/1894) and added the link in comment